internal/modproxy/proxy.go

```diff
@@ -0,0 +1,177 @@
+// configured on p.
+func (p Proxy) GetModFile(mod, version string) (*modfile.File, error) {
+	for _, proxy := range p.proxies {
+		u := proxy + "/" + path.Join(mod, "@v", semver.Canonical(version)+".mod")
+		resp, err := p.g.Get(u)
+		if err != nil {
+			return nil, fmt.Errorf("error fetching module versions from %s: %w", u, err)
+		}
+		defer func() {
+			if resp.Body != nil {
+				_ = resp.Body.Close()
+			}
+		}()
+		switch resp.StatusCode {
+		case http.StatusOK:
+			data, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("error reading the module proxy respons from %s: %w", u, err)
+			}
+			f, err := modfile.ParseLax(mod+"@"+version+"/go.mod", data, nil)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing go.mod from %s: %w", u, err)
+			}
+			return f, nil
+		case http.StatusNotFound, http.StatusGone:
+			// try the next proxy
+			continue
+		default:
+			return nil, fmt.Errorf("unexpected response code from %s: %s", u, resp.Status)
+		}
+	}
+	return nil, fmt.Errorf("the specified module was not found")
+}
+
+// Getter defines a type, such as http.Client, that can perform an HTTP GET request and return
+// the result.
+//
+// This interface is defined so that consumers and tests can provide potentially customized implementations,
+// but http.DefaultClient (or some other constructed http.Client instance) will likely be the most
+// common implementation used.
+type Getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+// GetCurrentVersion returns the highest known version of the specified module, as returned by the
+// system's module proxy.
+func GetCurrentVersion(g Getter, mod string, includePrerelease bool) (string, error) {
+	p := NewFromEnv(g)
+	return p.GetCurrentVersion(mod, includePrerelease)
+}
+
+// GetModuleVersions uses the provided getter instance to retrieve a list of module versions for the
+// specified module by querying the system Go module proxy ($GOPROXY)
+func GetModuleVersions(g Getter, mod string) ([]string, error) {
+	p := NewFromEnv(g)
+	return p.GetModuleVersions(mod)
+}
+
+// GetModFile uses the provided getter instance to retrieve the go.mod file for the specified module
+// by querying the system Go module proxy ($GOPROXY)
+func GetModFile(g Getter, mod, version string) (*modfile.File, error) {
+	p := NewFromEnv(g)
+	return p.GetModFile(mod, version)
+}
+
+// getModProxies returns a list of Go module proxies by parsing the GOPROXY environment variable.  If
+// no proxy is set ($GOPROXY is unset or "") this function returns a single result containing the
+// Google public proxy.
+func getModProxies() []string {
+	ev := os.Getenv("GOPROXY")
+	if ev == "" {
+		return []string{"https://proxy.golang.org"}
+	}
+	var results []string
+	for _, s := range strings.Split(ev, ",") {
+		// not trying to deal with pulling go.mod directly from various VCSs for now
+		if s != "direct" {
+			results = append(results, s)
```

Comment:
Since all usages above of `proxy` assume no trailing slash, should we trim here to avoid double slashes when building the path?

Reply:
Good point.  I was originally using `path.Join()` to ensure there was a consistent slash but that function also converts the double-slash after the URL scheme to a single one so I switched to manually concatenating strings.